test/sim/moves/curse.js

```diff
@@ -10,82 +10,135 @@ describe('Curse', function () {
+		battle.makeChoices();
+		assert.equal(greninja.hp, greninja.maxhp - Math.floor(greninja.maxhp / 2) - curseResidual, `Greninja should have Cursed itself`);
+		assert.fullHP(caterpie);
 
 		battle.makeChoices('move spite', 'auto');
-		assert.notEqual(hps[0], battle.p1.active[0].hp); // Curse user is hurt by residual damage
-		assert.equal(hps[1], battle.p2.active[0].hp); // Foe unaffected
+		assert.equal(greninja.hp, greninja.maxhp - Math.floor(greninja.maxhp / 2) - curseResidual * 2, `Greninja should have taken Curse damage again`);
+		assert.fullHP(caterpie);
 	});
 
-	it('should curse the target if a Ghost user has Protean', function () {
-		battle = common.createBattle();
-		battle.setPlayer('p1', {team: [{species: "Gengar", ability: 'protean', item: '', moves: ['curse', 'spite']}]});
-		battle.setPlayer('p2', {team: [{species: "Caterpie", ability: 'shedskin', item: '', moves: ['stringshot']}]});
+	it(`should curse the target if a Ghost user has Protean`, function () {
+		battle = common.createBattle([[
+			{species: 'Gengar', ability: 'protean', moves: ['curse']},
+		], [
+			{species: 'Caterpie', moves: ['sleeptalk']},
+		]]);
+		const gengar = battle.p1.active[0];
+		const caterpie = battle.p2.active[0];
+		const curseResidual = Math.floor(caterpie.maxhp / 4);
+		battle.makeChoices();
+		assert.equal(gengar.hp, gengar.maxhp - Math.floor(gengar.maxhp / 2));
+		assert.equal(caterpie.hp, caterpie.maxhp - curseResidual);
+
+		battle.makeChoices();
+		assert.equal(gengar.hp, gengar.maxhp - Math.floor(gengar.maxhp / 2));
+		assert.equal(caterpie.hp, caterpie.maxhp - curseResidual * 2);
+	});
 
-		battle.makeChoices('auto', 'auto');
-		const hps = [battle.p1.active[0].hp, battle.p2.active[0].hp];
-		assert.notEqual(hps[0], battle.p1.active[0].maxhp); // Curse user cut its HP down
-		assert.notEqual(hps[1], battle.p2.active[0].maxhp); // Curse residual damage
+	it(`should target either random opponent if the target is an ally`, function () {
+		battle = common.createBattle({gameType: 'doubles'}, [[
+			{species: 'Wynaut', moves: ['sleeptalk']},
+			{species: 'Gengar', moves: ['curse']},
+		], [
+			{species: 'Caterpie', moves: ['sleeptalk']},
+			{species: 'Metapod', moves: ['sleeptalk']},
+		]]);
+		battle.makeChoices('move sleeptalk, move curse -1', 'auto');
+
+		const wynaut = battle.p1.active[0];
+		const caterpie = battle.p2.active[0];
+		const metapod = battle.p2.active[1];
+		assert.fullHP(wynaut);
+		assert(caterpie.maxhp !== caterpie.hp || metapod.maxhp !== metapod.hp, `Either Caterpie or Metapod should have lost HP from Curse`);
+	});
 
-		battle.makeChoices('move spite', 'auto');
-		// Check residual damage
-		assert.equal(hps[0], battle.p1.active[0].hp); // Curse user unaffected
-		assert.notEqual(hps[1], battle.p2.active[0].hp); // Curse residual damage
+	it(`[Gen 7] should target the ally if the target is an ally`, function () {
+		battle = common.gen(7).createBattle({gameType: 'doubles'}, [[
```

Comment:
If you're reverting the behaviour in the Gen 8 mod then surely that's the one you should use for testing?

Reply:
(Or should the behaviour actually have been reverted in the Gen 7 mod?)

Reply:
Gen 8 Curse can't target an ally in the first place. I took off the inheritance in the Gen 8 mod because I think it makes more sense to drop it there rather than 7 in case there's some jank with LightningRod Electrify Curse targeting or something that I haven't tested in 8.